enterprise/server/util/ocicache/ocicache.go

```diff
@@ -38,8 +45,161 @@ const (
+// Returns empty string if not found or expired.
+// Automatically removes expired entries.
+func (r *tagResolver) Get(ctx context.Context, tag string) string {
+	if r == nil || r.lru == nil {
+		return ""
+	}
+
+	r.mu.RLock()
+	entry, ok := r.lru.Get(tag)
+	r.mu.RUnlock()
+
+	if !ok {
+		return ""
+	}
+
+	// Check expiration using the stored clock
+	if r.clock.Now().After(entry.expiration) {
+		// Expired - remove it
+		r.mu.Lock()
+		r.lru.Remove(tag)
+		r.mu.Unlock()
+		return ""
+	}
+
+	return entry.digest
+}
+
+// Set stores a digest for the given tag with automatic expiration.
+func (r *tagResolver) Set(ctx context.Context, tag string, digest string) {
+	if r == nil || r.lru == nil {
+		return
+	}
+
+	entry := tagToDigestEntry{
+		digest:     digest,
+		expiration: r.clock.Now().Add(tagToDigestTTL),
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.lru.Add(tag, entry)
+}
+
+// ReadThroughFetcher is an interface for fetching OCI blobs and manifests from an upstream registry.
+// Implementations may tee data to/from a cache while fetching from the upstream.
+type ReadThroughFetcher interface {
+	// Head makes a HEAD request to fetch metadata about a manifest.
+	Head(ctx context.Context, ref gcrname.Reference) (*gcr.Descriptor, error)
+	// Get fetches a manifest from the upstream registry.
+	//TODO(dan) Get(ctx context.Context, ref gcrname.Reference) (*remote.Descriptor, error)
+	// Layer fetches a layer (blob) from the upstream registry.
+	//TODO(dan) Layer(ctx context.Context, ref gcrname.Digest) (gcr.Layer, error)
+	// LayerFromDescriptor fetches a layer with a known descriptor, avoiding extra HTTP requests.
+	// The parentImage is used to look up DiffID from the image config without fetching the layer blob.
+	//TODO(dan) LayerFromDescriptor(ctx context.Context, ref gcrname.Digest, desc gcr.Descriptor, parentImage gcr.Image) (gcr.Layer, error)
+}
+
+// readThroughFetcher implements ReadThroughFetcher by wrapping a remote.Puller.
+type readThroughFetcher struct {
+	puller      *remote.Puller
+	acClient    repb.ActionCacheClient
+	bsClient    bspb.ByteStreamClient
+	tagResolver interfaces.OCITagResolver
+	useCache    bool
+}
+
+// NewReadThroughFetcher creates a new ReadThroughFetcher that fetches OCI resources from an upstream registry
+// and tees data to/from the cache.
+// If useCache is false, caching is disabled and all requests go directly to the upstream registry.
+func NewReadThroughFetcher(env environment.Env, useCache bool, opts ...remote.Option) (ReadThroughFetcher, error) {
+	puller, err := remote.NewPuller(opts...)
+	if err != nil {
+		return nil, err
+	}
+	return &readThroughFetcher{
+		puller:      puller,
+		acClient:    env.GetActionCacheClient(),
+		bsClient:    env.GetByteStreamClient(),
+		tagResolver: env.GetOCITagResolver(),
+		useCache:    useCache,
+	}, nil
+}
+
+func (c *readThroughFetcher) Head(ctx context.Context, ref gcrname.Reference) (*gcr.Descriptor, error) {
```

Comment:
d'oh, sorry - I guess when I suggested the `OCITagResolver` naming I was picturing that the `OCITagResolver` would also be the thing doing the head request, so it would "resolve" the tag and not just be an in-memory cache. Would it be cleaner to do it that way, or does it complicate things because we need the puller instance per-request to manage auth for us? (or, would it make sense to have the resolver do the HEAD request using plain old HTTP + auth headers, and maybe call this thing `ResolveTag`? and then at that point, would it make sense to just use the resolver directly instead of the readThroughFetcher, at least for the tag resolution part? Sorry for all the questions :sweat_smile: )